MIG-834: More PV values to edit in MigPlan (#38095)
openshift/openshift-docs: modules/migration-editing-pvs-in-migplan.adoc

[id="migration-editing-pvs-in-migplan_{context}"]
= Editing persistent volume attributes

After you create a `MigPlan` custom resource (CR), the `MigrationController` CR discovers the persistent volumes (PVs). The `spec.persistentVolumes` block and the `status.destStorageClasses` block are added to the `MigPlan` CR.

You can edit the values in the `spec.persistentVolumes.selection` block. If you change values outside this block, they are overwritten when the `MigPlan` CR is reconciled by the `MigrationController` CR.

[NOTE]
====
If a value for the `spec.persistentVolumes.selection.storageClass` parameter is not specified, a default storage class on the destination cluster is chosen according to the following logic:

> 

Author: Thanks. I didn't realize that the default was already filled in or that the value could be null. Text updated.

[id="migration-editing-pvs-in-migplan_{context}"]
= Editing persistent volume attributes

After you create a `MigPlan` custom resource (CR), the `MigrationController` CR discovers the persistent volumes (PVs). The `spec.persistentVolumes` block and the `status.destStorageClasses` block are added to the `MigPlan` CR.

> You can edit the values in the `spec.persistentVolumes.selection` block. If you change values outside this block, they are overwritten when the `MigPlan` CR is reconciled by the `MigrationController` CR.

Reviewer: ```suggestion
You can edit the values in the `spec.persistentVolumes.selection` block. If you change values outside of the `spec.persistentVolumes.selection` block, the values are overwritten when the `MigPlan` CR is reconciled by the `MigrationController` CR.
```

Suggestion for clearer language.

Author: "outside of" is not good usage because "outside" does not take that preposition. "outside the box", not "outside of the box". However, the other changes are OK.

[NOTE]
====
The default value for the `spec.persistentVolumes.selection.storageClass` parameter is determined by the following logic:

> . If the source cluster PV is Gluster or NFS, the default is either `cephfs`, for `accessMode: ReadWriteMany`, or `cephrbd`, for `accessMode: ReadWriteOnce`.

Reviewer 2: @apinnick I don't understand it 
> If **the source cluster** PV is Gluster or NFS, the **default** is either `cephfs ..

Author: It is in Scott's PR: https://github.com/konveyor/mig-operator/pull/767/files

docs/usage/MigPlanPVSelection.md, line 44

Author: @sseago Can you confirm this? Will also ping you in Slack.

Reviewer 3: @XinRedhat The purpose behind this was to facilitate OCS users. For 3 clusters (either with OCS, using gluster, or without,  using NFS), the intent is to default to OCS4 ceph storage classes *if* they're available. If they're not available, we can't default to them, so we fall back to the other defaults (same provisioner if available, otherwise the default storageclass for the dest/target cluster).

. If the PV is neither Gluster nor NFS _or_ if `cephfs` or `cephrbd` are not available, the default is a storage class for the same provisioner.
. If a storage class for the same provisioner is not available, the default is the default storage class of the destination cluster.

You can change the `storageClass` value to the value of any `name` parameter in the `status.destStorageClasses` block of the `MigPlan` CR.

> If the `storageClass` value is `null`, the PV will have no storage class after migration. This option is appropriate if, for example, you want to move the PV to an NFS volume on the destination cluster.

Reviewer 2: >If the `storageClass` value is `null`
I think the `storageClass` value is "", an empty string

      name: mysql
      namespace: mysql-persistent
    selection:
      action: <copy> <1>
      copyMethod: <filesystem> <2>
      verify: <true> <3>
      storageClass: <gp2> <4>
      accessMode: <ReadWriteMany> <5>
    storageClass: cephfs
----
<1> Allowed values are `move`, `copy`, and `skip`. If only one action is supported, the default value is the supported action. If multiple actions are supported, the default value is `copy`.
<2> Allowed values are `snapshot` and `filesystem`. Default value is `filesystem`.
<3> For file system copy method only: Set to `true` to verify the files after they are copied. Default value is `null`.

> <4> Storage class on the destination cluster. If the source cluster PV is Gluster or NFS, the default value is `cephfs` for `ReadWriteMany` or `cephrbd` for `ReadWriteOnce`, if they are available. If `cephfs` or `cephrbd` are not available or if the source cluster PV is a different type, the default is the storage class _for the same provisioner_, if it is available. If not, the default is the default storage class of the destination cluster.

Reviewer 3: Storage classes available are those listed in `status.destStorageClasses`. The `name` from any listed storage class is a valid value for `storageClass` here.

Author: Are you saying that the `name` field will be populated after PV discovery? The user does not need to know it in advance?

Reviewer 3: The entire status.destStorageClasses struct is populated automatically -- that's why we've put it in status, as there's nothing user-editable there. Basically the user selection for `selection.storageClass`, if it's not empty, must correspond to the "name" field from one of the destStorageClasses.

Author: Great. That makes things much clearer.

      name: mysql
      namespace: mysql-persistent
    selection:
      action: <copy> <1>
      copyMethod: <filesystem> <2>
      verify: <true> <3>
      storageClass: <gp2> <4>
      accessMode: <ReadWriteMany> <5>
    storageClass: cephfs
----
<1> Allowed values are `move`, `copy`, and `skip`. If only one action is supported, the default value is the supported action. If multiple actions are supported, the default value is `copy`.
<2> Allowed values are `snapshot` and `filesystem`. Default value is `filesystem`.
<3> For file system copy method only: Set to `true` to verify the files after they are copied. Default value is `null`.

> <4> Allowed values are listed in the `status.destStorageClasses` block of the `MigPlan` CR. The value must match the `name` field of one of the `status.destStorageClasses` values. If this value is not specified, a default storage class is selected.

Reviewer 3: An empty value is also valid here. If a value is not specified, then the resulting destination PVC will have no storage class specified.

      name: mysql
      namespace: mysql-persistent
    selection:
      action: <copy> <1>
      copyMethod: <filesystem> <2>
      verify: true <3>
      storageClass: <gp2> <4>
      accessMode: <ReadWriteMany> <5>
    storageClass: cephfs
----

> <1> Allowed values are `move`, `copy`, and `skip`. If only one action is supported, the default value is the supported action. If multiple actions are supported, the default value is `copy`.

Reviewer: ```suggestion
<1> Allowed values are `move`, `copy`, and `skip`. If only one action is supported, the default value is the supported action. If multiple actions are supported, the default value is `copy`.
```
Allowed or Accepted?  Maybe change to accepted or supported values? 

I know per the IBM style guide, we shouldn't use the word "allow" but I'm unsure about "allowed values". 

> Avoid` stating that inanimate objects grant abilities to people, as in "the product allows you to...". Whenever possible, use a direct, user-focused alternative such as "you can use the product to..." or "with this product, users can..."

Author: "Allowed values" occurs 21 times in OCP docs. "Accepted values" occurs 3 times.

Author: The IBM style guide quote does not apply because we are using "allowed" as an adjective, not a verb. "Permitted values" would also work but it is never used in OCP docs.
